File: android/src/main/java/com/faceverify/FaceVerifyPlugin.kt
package com.faceverify

import android.media.Image
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.mrousavy.camera.frameprocessors.Frame
import com.mrousavy.camera.frameprocessors.FrameProcessorPlugin
import java.nio.ByteBuffer
import kotlin.math.abs
import kotlin.math.sqrt

class FaceVerifyPlugin : FrameProcessorPlugin() {

  private val faceDetector = FaceDetection.getClient(
    FaceDetectorOptions.Builder()
      .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
      .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
      .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
      .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
      .build()
  )

  private fun orientationDegrees(frame: Frame): Int {
    val name = frame.orientation.toString().uppercase()
    return when {
      name.contains("LANDSCAPE_LEFT")        -> 90
      name.contains("PORTRAIT_UPSIDE_DOWN")  -> 180
      name.contains("LANDSCAPE_RIGHT")       -> 270
      else                                   -> 0
    }
  }

  /**
   * Computes Laplacian variance on the Y (luma) plane of the frame.
   * Samples every 8th pixel for performance. Higher = sharper.
   *
   * IMPORTANT: All returned numbers must be Double — JSI cannot convert Float.
   */
  private fun laplacianVariance(image: Image): Double {
    val plane = image.planes[0]          // Y plane
    val buffer: ByteBuffer = plane.buffer
    val rowStride = plane.rowStride
    val width = image.width
    val height = image.height
    val step = 8

    val values = mutableListOf<Double>()

    for (y in 1 until height - 1 step step) {
      for (x in 1 until width - 1 step step) {
        val center = (buffer.get(y * rowStride + x).toInt() and 0xFF).toDouble()
        val top    = (buffer.get((y - 1) * rowStride + x).toInt() and 0xFF).toDouble()
        val bottom = (buffer.get((y + 1) * rowStride + x).toInt() and 0xFF).toDouble()
        val left   = (buffer.get(y * rowStride + (x - 1)).toInt() and 0xFF).toDouble()
        val right  = (buffer.get(y * rowStride + (x + 1)).toInt() and 0xFF).toDouble()
        values.add(abs(top + bottom + left + right - 4.0 * center))
      }
    }

    if (values.isEmpty()) return 0.0
    val mean = values.average()
    return values.map { (it - mean) * (it - mean) }.average()
  }

  override fun callback(frame: Frame, arguments: Map<String, Any>?): Any {
    val mediaImage: Image = frame.image
    val sharpness = laplacianVariance(mediaImage)
    val inputImage = InputImage.fromMediaImage(mediaImage, orientationDegrees(frame))

    return try {
      val faces = Tasks.await(faceDetector.process(inputImage))

      if (faces.isEmpty()) {
        return mapOf("detected" to false, "sharpness" to sharpness)
      }

      val face = faces.first()
      val box = face.boundingBox

      // All numeric values must be Double — JSI cannot convert Java Float.
      mapOf(
        "detected" to true,
        "bounds" to mapOf(
          "x"      to box.left.toDouble(),
          "y"      to box.top.toDouble(),
          "width"  to box.width().toDouble(),
          "height" to box.height().toDouble()
        ),
        "frameWidth"              to mediaImage.width.toDouble(),
        "frameHeight"             to mediaImage.height.toDouble(),
        "yawAngle"                to face.headEulerAngleY.toDouble(),
        "pitchAngle"              to face.headEulerAngleX.toDouble(),
        "rollAngle"               to face.headEulerAngleZ.toDouble(),
        "leftEyeOpenProbability"  to (face.leftEyeOpenProbability?.toDouble()  ?: -1.0),
        "rightEyeOpenProbability" to (face.rightEyeOpenProbability?.toDouble() ?: -1.0),
        "sharpness"               to sharpness
      )
    } catch (e: Exception) {
      mapOf("detected" to false, "sharpness" to sharpness)
    }
  }
}

File: ios/FaceVerifyPlugin.swift
import VisionCamera
import MLKitFaceDetection
import MLKitVision
import UIKit
import AVFoundation
import Accelerate

@objc(FaceVerifyPlugin)
public class FaceVerifyPlugin: FrameProcessorPlugin {

  private var faceDetector: FaceDetector

  public override init(proxy: VisionCameraProxyHolder, options: [AnyHashable: Any]? = [:]) {
    let opts = FaceDetectorOptions()
    opts.performanceMode = .fast
    opts.classificationMode = .all
    opts.landmarkMode = .none
    opts.contourMode = .none
    opts.isTrackingEnabled = false
    self.faceDetector = FaceDetector.faceDetector(options: opts)
    super.init(proxy: proxy, options: options)
  }

  public override func callback(
    _ frame: Frame,
    withArguments arguments: [AnyHashable: Any]?
  ) -> Any {
    let sharpness = laplacianVariance(frame.buffer)

    let image = VisionImage(buffer: frame.buffer)
    image.orientation = uiOrientation(from: frame)

    do {
      let faces = try faceDetector.results(in: image)
      guard let face = faces.first else {
        return ["detected": false, "sharpness": sharpness]
      }

      let bufW = Double(CVPixelBufferGetWidth(frame.buffer))
      let bufH = Double(CVPixelBufferGetHeight(frame.buffer))

      return [
        "detected": true,
        "bounds": [
          "x": Double(face.frame.origin.x),
          "y": Double(face.frame.origin.y),
          "width": Double(face.frame.size.width),
          "height": Double(face.frame.size.height),
        ],
        "frameWidth": bufW,
        "frameHeight": bufH,
        "yawAngle": Double(face.headEulerAngleY),
        "pitchAngle": Double(face.headEulerAngleX),
        "rollAngle": Double(face.headEulerAngleZ),
        "leftEyeOpenProbability": Double(face.leftEyeOpenProbability),
        "rightEyeOpenProbability": Double(face.rightEyeOpenProbability),
        "sharpness": sharpness,
      ]
    } catch {
      return ["detected": false, "sharpness": sharpness]
    }
  }

  // MARK: - Sharpness (Laplacian variance)

  /// Computes the Laplacian variance of the luma channel as a sharpness metric.
  /// Samples every 8th pixel for performance. Higher = sharper.
  private func laplacianVariance(_ buffer: CVPixelBuffer) -> Double {
    CVPixelBufferLockBaseAddress(buffer, .readOnly)
    defer { CVPixelBufferUnlockBaseAddress(buffer, .readOnly) }

    let width = CVPixelBufferGetWidth(buffer)
    let height = CVPixelBufferGetHeight(buffer)
    let bytesPerRow = CVPixelBufferGetBytesPerRow(buffer)
    guard let base = CVPixelBufferGetBaseAddress(buffer) else { return 0 }

    let pixels = base.assumingMemoryBound(to: UInt8.self)
    let step = 8
    var values: [Double] = []
    values.reserveCapacity(((width / step) - 2) * ((height / step) - 2))

    // BGRA layout — channel index 1 (G) is a reasonable luma approximation
    // without needing to know the exact pixel format.
    let channel = 1

    for y in stride(from: 1, to: height - 1, by: step) {
      for x in stride(from: 1, to: width - 1, by: step) {
        let center = Int(pixels[y * bytesPerRow + x * 4 + channel])
        let top    = Int(pixels[(y - 1) * bytesPerRow + x * 4 + channel])
        let bottom = Int(pixels[(y + 1) * bytesPerRow + x * 4 + channel])
        let left   = Int(pixels[y * bytesPerRow + (x - 1) * 4 + channel])
        let right  = Int(pixels[y * bytesPerRow + (x + 1) * 4 + channel])
        let lap = Double(abs(top + bottom + left + right - 4 * center))
        values.append(lap)
      }
    }

    guard !values.isEmpty else { return 0 }
    let mean = values.reduce(0, +) / Double(values.count)
    let variance = values.map { ($0 - mean) * ($0 - mean) }.reduce(0, +) / Double(values.count)
    return variance
  }

  // MARK: - Orientation helper

  private func uiOrientation(from frame: Frame) -> UIImage.Orientation {
    switch frame.orientation {
    case .portrait:           return .right
    case .portraitUpsideDown: return .left
    case .landscapeLeft:      return .up
    case .landscapeRight:     return .down
    @unknown default:         return .right
    }
  }
}
